### ISSUE TITLE
Adding Special Tokens to GPT2 doesn't have any effect

### ISSUE THREAD

**Issue by souravrai98** [huggingface/transformers#31141]
### System Info

- `transformers` version: 4.40.2
- Platform: Linux-5.4.0-167-generic-x86_64-with-glibc2.31
- Python version: 3.12.3
- Huggingface_hub version: 0.23.0
- Safetensors version: 0.4.3
- Accelerate version: 0.30.1
- Accelerate config:    not found
- PyTorch version (GPU?): 2.3.0 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: Yes
- Using distributed or parallel set-up in script?: No

### Who can help?

@Arthur @ArthurZucker 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

```
import logging
import time
from transformers import GPT2Tokenizer

logger = logging.getLogger(__name__)

def tokenize_dataset_with_target(data, input_length=10, target_length=10):
    logger.info("--------------Tokenization Started---------------------")
    start_time = time.time()

    tokenizer = GPT2Tokenizer.from_pretrained('gpt2')

    # Get the number of columns in the data
    num_columns = data.shape[1]
    
    # Add row and column tokens to the tokenizer
    special_tokens = {'additional_special_tokens': ['<|target|>', '<row>'] + [f'<column{i}>' for i in range(1, num_columns + 1)]}
    tokenizer.add_special_tokens(special_tokens)
    logger.info("Added special tokens to the tokenizer.")
    logger.info("Special tokens: %s", tokenizer.special_tokens_map)
    
    # Print the token ID for the <|target|> token
    target_token_id = tokenizer.convert_tokens_to_ids('<|target|>')
    logger.info("Token ID for '<|target|>': %d", target_token_id)
    print(f"Token ID for '<|target|>': {target_token_id}")

    # Print the token ID for the <row> token
    row_token_id = tokenizer.convert_tokens_to_ids('<row>')
    logger.info("Token ID for '<row>': %d", row_token_id)
    print(f"Token ID for '<row>': {row_token_id}")

    # Print token IDs for all column tokens
    for i in range(1, num_columns + 1):
        column_token = f'<column{i}>'
        column_token_id = tokenizer.convert_tokens_to_ids(column_token)
        logger.info("Token ID for '%s': %d", column_token, column_token_id)
        print(f"Token ID for '{column_token}': {column_token_id}")

    tokenizer.pad_token = tokenizer.eos_token

    sequences = []

    for i in range(len(data) - input_length - target_length + 1):
        input_rows = data.iloc[i:i + input_length]
        target_rows_series = data.iloc[i + input_length:i + input_length + target_length]

        # Join each row with its respective column tokens
        input_text = ' '.join(['<row> ' + ' '.join([f'<column{j+1}> {cell}' for j, cell in enumerate(row)]) for row in input_rows.values])
        target_text = ' '.join(['<row> ' + ' '.join([f'<column{j+1}> {cell}' for j, cell in enumerate(row)]) for row in target_rows_series.values])

        combined_text = input_text + ' <|target|> ' + target_text
        print(f"Combined text: {combined_text}")  # Original text for DEBUG
        print(f"Combined text length: {len(combined_text.split())}")

        tokenized_text = tokenizer(combined_text, return_tensors='pt', padding='max_length', max_length=512, truncation=True)
        
        # Debugging step to check <|target|> token presence
        token_ids = tokenized_text['input_ids'].squeeze().tolist()
        if target_token_id not in token_ids:
            logger.info(f"Error: <|target|> token {target_token_id} not found in token_ids")
            print(f"Token IDs: {token_ids}")
            print(f"Decoded tokens: {tokenizer.decode(token_ids)}")  # Decode to see why it might be missing


        sequences.append(tokenized_text)

    end_time = time.time()
    elapsed_time = end_time - start_time
    logger.info("----------------------Tokenization Ended----------------")
    logger.info(f"Tokenization completed in {elapsed_time:.2f} seconds.")    

    return sequences, tokenizer
```
```

from transformers import GPT2LMHeadModel
import torch
from torch.nn import CrossEntropyLoss, BCEWithLogitsLoss, Linear
from transformers.modeling_outputs import BaseModelOutputWithPastAndCrossAttentions, CausalLMOutputWithCrossAttentions

class TabularModel(GPT2LMHeadModel):
    def __init__(self, config, vocab, tokenizer):
        super().__init__(config)
        self.vocab = vocab
        self.bce_loss = BCEWithLogitsLoss()  # Define BCE Loss here
        self.tokenizer = tokenizer
        
        self.binary_classification_head = Linear(config.hidden_size, 1)
    def forward(
            self,
            input_ids=None,
            past_key_values=None,
            attention_mask=None,
            token_type_ids=None,
            position_ids=None,
            head_mask=None,
            inputs_embeds=None,
            labels=None,
            binary_labels=None,  # Add parameter for binary labels
            use_cache=True,
            return_dict=None,
            output_attentions = None,
            output_hidden_states = None,
    ):
        
        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
        transformer_outputs = self.transformer(
            input_ids,
            past_key_values=past_key_values,
            attention_mask=attention_mask,
            token_type_ids=token_type_ids,
            position_ids=position_ids,
            head_mask=head_mask,
            inputs_embeds=inputs_embeds,
            use_cache=use_cache,
            output_attentions=output_attentions,
            output_hidden_states=output_hidden_states,
            return_dict=return_dict,
        )
        hidden_states = transformer_outputs[0]
        lm_logits = self.lm_head(hidden_states)

        outputs = (lm_logits,) + transformer_outputs[1:]
        total_loss = 0.0  
        
        if labels is not None:
            # Identify the <|target|> token positions
            target_token_id = self.tokenizer.convert_tokens_to_ids('<|target|>')
            target_token_indices = (input_ids == target_token_id).nonzero(as_tuple=False)

            if target_token_indices.size(0) > 0:
                for target_token_index in target_token_indices.tolist():
                    batch_idx = target_token_index[0]
                    token_idx = target_token_index[1]

                    # Determine row boundaries using <row> tokens
                    row_token_id = self.tokenizer.convert_tokens_to_ids('<row>')
                    row_token_indices = (input_ids[batch_idx] == row_token_id).nonzero(as_tuple=False)

                    # Iterate through rows after the <|target|>
                    row_indices = row_token_indices[(row_token_indices[:, 0] > token_idx).nonzero().squeeze()]

                    if row_indices.numel() > 0:
                        for idx in range(len(row_indices)):
                            start_idx = row_indices[idx].item()
                            end_idx = None
                            if idx + 1 < len(row_indices):
                                end_idx = row_indices[idx + 1].item()

                            # Compute Cross Entropy Loss for tokens in this row segment
                            shift_labels = labels[batch_idx, start_idx:end_idx].contiguous()
                            shift_logits = lm_logits[batch_idx, start_idx:end_idx, :].contiguous()

                            loss_fct = CrossEntropyLoss(ignore_index=self.tokenizer.eos_token_id)
                            ce_loss = loss_fct(shift_logits.view(-1, shift_logits.size(-1)), shift_labels.view(-1))

                            # Avoid in-place operation by creating a new tensor for total_loss
                            total_loss = total_loss + ce_loss


            outputs = (total_loss,) + outputs                
                            
        if return_dict:
            return CausalLMOutputWithCrossAttentions(
            loss=total_loss,
            logits=lm_logits,
            past_key_values=transformer_outputs.past_key_values,
            hidden_states=transformer_outputs.hidden_states,
            attentions=transformer_outputs.attentions,
            cross_attentions=transformer_outputs.cross_attentions,
        )
        
  #      print(f"Total Loss: {total_loss.item()}")
        return outputs
```

```
import torch
from transformers import GPT2LMHeadModel, GPT2Tokenizer, GPT2Config
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from model import TabFormerGPT2LMHeadModel
from collections import Counter
import transformers
import random

def load_model_and_tokenizer(model_dir, tokenizer_dir, vocab):
    
    #config = transformers.GPT2Config()
    tokenizer = GPT2Tokenizer.from_pretrained(tokenizer_dir)
    
    config = GPT2Config.from_pretrained(model_dir)
    model = TabFormerGPT2LMHeadModel.from_pretrained(model_dir, config=config, vocab=vocab, tokenizer=tokenizer)
    return model, tokenizer

def generate_sequences(model, tokenizer, prompt, max_length=50, num_return_sequences=1):
    inputs = tokenizer(prompt, return_tensors='pt').to(model.device)
    
    # Generate sequences
    output_sequences = model.generate(
        input_ids=inputs.input_ids,
        max_length=max_length,
        num_return_sequences=num_return_sequences,
        pad_token_id=tokenizer.eos_token_id,  # Ensure padding tokens are handled correctly
        top_k=40,  # Top-K sampling
        top_p=0.85,  # Top-p (nucleus) sampling
        temperature=0.7,  # Adjust temperature to control randomness
        do_sample=True,  # Enable sampling
    )
    
    return [tokenizer.decode(seq, skip_special_tokens=True) for seq in output_sequences]

def build_vocab(data, additional_tokens=['<|target|>', '<row>']):
    # Get the number of columns in the data excluding the 'comportement' column
    num_columns = data.shape[1] - 1

    # Add column tokens dynamically based on the number of columns
    column_tokens = [f'<column{i}>' for i in range(1, num_columns + 1)]
    all_additional_tokens = additional_tokens + column_tokens
    
    # Flatten the dataset (excluding 'comportement' column) into a list of words
    words = data.drop('comportement', axis=1).applymap(str).values.flatten().tolist()
    
    # Get the unique labels from the 'comportement' column
    labels = data['comportement'].unique().tolist()
    
    # Combine words, labels, and additional tokens to create a unified vocabulary
    combined_vocab = words + labels + all_additional_tokens
    
    # Use Counter to get the frequency of each token
    vocab_counter = Counter(combined_vocab)
    
    # Create a vocabulary dictionary with token: index mapping
    vocab = {word: i for i, (word, _) in enumerate(vocab_counter.items())}
    
    return vocab


class WrapperModel:
    def __init__(self, model):
        self.model = model

    def generate(self, *args, **kwargs):
        # Remove return_dict_in_generate if it exists
        kwargs.pop('return_dict_in_generate', None)
        return self.model.generate(*args, **kwargs)
    
    def __getattr__(self, name):
        return getattr(self.model, name)


# Paths to the fine-tuned model and tokenizer
model_dir = './results1/checkpoint-50000'  # Adjust this path to where your fine-tuned model is saved
tokenizer_dir = './tokenizer'  # Adjust this path to where your tokenizer is saved

# Load the model and tokenizer



file_path = "adxsrv_label_clean_v3.csv"
data = pd.read_csv(file_path)
data = data.drop(columns=['appname', 'gid', 'label'])
total_rows = data.shape[0]

# Initialize LabelEncoder and encode 'comportement' column
label_encoder = LabelEncoder()
data['comportement'] = label_encoder.fit_transform(data['comportement'])

vocab = build_vocab(data)


model, tokenizer = load_model_and_tokenizer(model_dir, tokenizer_dir, vocab)

model = WrapperModel(model)
# Move the model to GPU if available
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
model = model.to(device)

n = random.randrange(1,total_rows)
# Prepare a prompt including 'comportement' column
prompt_data = data.iloc[n].tolist()

# Add 10 to the first five columns
for i in range(min(5, len(prompt_data))):
    prompt_data[i] += 10

# Convert to string while preserving the special tokens structure
formatted_prompt = ' '.join(
    [f'<row>'] + [f'<column{i+1}> {cell}' for i, cell in enumerate(prompt_data)] + ['<|target|>']
)

print(f"Formatted Prompt: {formatted_prompt}")


# Tokenize and convert to tensor
tokenized_prompt = tokenizer(formatted_prompt, return_tensors='pt').input_ids.to(device)

print(f"Tokenized Prompt IDs: {tokenized_prompt.tolist()}")

# Define generation parameters
max_length = 256
num_return_sequences = 3



# Greedy Search
greedy_output = model.generate(
    input_ids=tokenized_prompt,
    max_length=max_length,
   
    pad_token_id=tokenizer.eos_token_id,
    do_sample=False
)


# Print out the sequences with decoded tokens
for i, output in enumerate(greedy_output):
    decoded_output = tokenizer.decode(output, skip_special_tokens=False)
    token_ids = output.tolist()
    print(f"\nGenerated Sequence {i + 1}:")
    print(decoded_output)
    print("Token IDs:", token_ids)
    
    
# Beam Search
beam_output = model.generate(
    input_ids=tokenized_prompt,
    max_length=max_length,
    num_return_sequences=num_return_sequences,
    pad_token_id=tokenizer.eos_token_id,
    num_beams=20,
    no_repeat_ngram_size=4, 
    early_stopping=True,
    do_sample=True
)

# Sampling
sampling_output = model.generate(
    input_ids=tokenized_prompt,
    max_length=max_length,
    pad_token_id=tokenizer.eos_token_id,
    do_sample=True,

    temperature=1.3
)

# Print the outputs for comparison
print("Actual Prompt Data: \n", prompt_data)
print("\nGreedy Search Output:")
for seq in greedy_output:
    print(tokenizer.decode(seq, skip_special_tokens=False))

print("\nBeam Search Output:")
for seq in beam_output:
    print(tokenizer.decode(seq, skip_special_tokens=False))

print("\nSampling Output:")
for seq in sampling_output:
    print(tokenizer.decode(seq, skip_special_tokens=False))
```

### Expected behavior

I am trying to make GPT2 learn the data that is stored in a table. So, for that purpose, I added special tokens to the tokenization process. It works something like this: every row starts with the <row> token. Then I have <column> token before the value of the column. I want to make the learning such that it sees 10 rows and predicts the loss just for the next 10 rows. Hence, I have the input_length and target_length parameters. The input rows and the output rows are separated by a new special token <target> token. I have also modified the forward function to predict just after the <target> token. But the expected generated output is never different. I always get the same generated output as the prompt. My prompt is the one of rows from my dataset. What might be the issue here? How do we rectify this issue of generating different outputs?

**Comment by LysandreJik** [comment]
Hello, thanks for opening an issue! We try to keep the github issues for bugs/feature requests.
Could you ask your question on the [forum](https://discuss.huggingface.co) instead?

Thanks!
